node/cmd/ccq/http.go

```diff
@@ -0,0 +1,175 @@
+package ccq
+
+import (
+	"crypto/ecdsa"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+	"time"
+
+	"github.com/certusone/wormhole/node/pkg/common"
+	gossipv1 "github.com/certusone/wormhole/node/pkg/proto/gossip/v1"
+	"github.com/certusone/wormhole/node/pkg/query"
+	"github.com/gorilla/mux"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+type queryRequest struct {
+	Bytes     string `json:"bytes"`
+	Signature string `json:"signature"`
+}
+
+type queryResponse struct {
+	Bytes      string   `json:"bytes"`
+	Signatures []string `json:"signatures"`
+}
+
+type httpServer struct {
+	topic            *pubsub.Topic
+	logger           *zap.Logger
+	env              common.Environment
+	permissions      Permissions
+	signerKey        *ecdsa.PrivateKey
+	pendingResponses *PendingResponses
+}
+
+func (s *httpServer) handleQuery(w http.ResponseWriter, r *http.Request) {
+	// Set CORS headers for all requests.
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+
+	// Set CORS headers for the preflight request
+	if r.Method == http.MethodOptions {
+
+		w.Header().Set("Access-Control-Allow-Methods", "PUT")
+		w.Header().Set("Access-Control-Allow-Headers", "Content-Type, X-Api-Key")
+		w.Header().Set("Access-Control-Max-Age", "3600")
+		w.WriteHeader(http.StatusNoContent)
+		return
+	}
+	var q queryRequest
+	err := json.NewDecoder(r.Body).Decode(&q)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// There should be one and only one API key in the header.
+	apiKey, exists := r.Header["X-Api-Key"]
+	if !exists || len(apiKey) != 1 {
+		s.logger.Debug("received a request without an api key", zap.Stringer("url", r.URL), zap.Error(err))
+		http.Error(w, "api key is missing", http.StatusBadRequest)
+		return
+	}
+
+	queryRequestBytes, err := hex.DecodeString(q.Bytes)
+	if err != nil {
+		s.logger.Debug("failed to decode request bytes", zap.Error(err))
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	signature, err := hex.DecodeString(q.Signature)
+	if err != nil {
+		s.logger.Debug("failed to decode signature bytes", zap.Error(err))
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	signedQueryRequest := &gossipv1.SignedQueryRequest{
+		QueryRequest: queryRequestBytes,
+		Signature:    signature,
+	}
+
+	if err := validateRequest(s.logger, s.env, s.permissions, s.signerKey, apiKey[0], signedQueryRequest); err != nil {
+		s.logger.Debug("invalid request", zap.String("api_key", apiKey[0]), zap.Error(err))
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	m := gossipv1.GossipMessage{
+		Message: &gossipv1.GossipMessage_SignedQueryRequest{
+			SignedQueryRequest: signedQueryRequest,
+		},
+	}
+
+	b, err := proto.Marshal(&m)
+	if err != nil {
+		s.logger.Debug("failed to marshal gossip message", zap.Error(err))
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
```

Comment:
I think that we should definitely log as `Error`, but maybe we can keep `http.StatusInternalServerError` since `validateRequest` will return `false` if we can't `Unmarshal` the `signedQueryRequest` - we should be able to `Marshal` `m` at this point (I think failure here would be a very rare occurrance)

Reply:
Oh totally agree here @kev1n-peters.